20221103-endpoint-authz.md

```diff
@@ -0,0 +1,304 @@
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Overview
+
+The endpoint authorization configuration extends the existing authentication config, allowing OpenFGA operators to require a specific subject, scope, or preshared key in order to access any given endpoint. To help reduce the verbosity of certain configurations we will also allow a top-level `global` authorization configuration that can be overridden on a per-endpoint basis.
+
+First, we must identify what kind of callers are allowed. This will differ depending on the authentication method; `presharedkey` or `oidc`. For `presharedkey` we use the key itself. For `oidc` we can support either an allowed list of scopes or an allowed list of subjects.
+
+Second, we must have a means to identify which routes need specific protections in the OpenFGA config. We must be careful not to use anything gRPC or HTTP specifc; this solution must work for either server configuration. The [`protobuf` API specification](https://github.com/openfga/api/blob/main/openfga/v1/openfga_service.proto#L14) is the common ground here, so let's use the RPC names (`ListStores`, `Read`, `Write`, `Check`, etc.) in the config.
+
+For simplicity, permissions granted on an endpoint will be a union of all allowed subjects or scopes. If an endpoint is not present in the authorization configuration and no global authorization is configured, then all authentiated clients will be allowed access.
+
+## Config structure
+The means for identifying an authorized client is different for `preshared` and `oidc`, so there will be a method-specific configuration object defined for both of them and nested under their respective configuration object.
+
+For both `preshared` and `oidc`, a key named `authz` will be added under the root. Under `authz` there will be two keys, `global` and `endpoints`. The `global` object will define the global authorization requirements, and the `endpoints` object will define the per-method overrides.
+
+The `endpoints` key will be an object, the keys of which are an enum with the defined `protobuf` RPC methods as allowed values. The values of the RPC method keys and the `global` keys will differ depending on the authentication method.
+
+`preshared` has one option, `keys`, which is a list of keys. These will be validated against the master list of allowed keys.
+
+`oidc` will have two options: `scopes` and `subjects`. Either or both can be present, and each value is a list of strings representing the allowed scopes or subjects, respectively.
+
+## Examples
+A preshared key config that authorizes one key globally, and different keys on `Write` and `CreateStore`:
+
+```yaml
+authn:
+  method: preshared
+  preshared:
+    keys:
+      - cool-key-1
+      - cool-key-2
+    authz:
```

Comment:
So that we're not mixing `authn` with `authz` I may recommend we introduce a separate block of config for the per-endpoint configuration. May I suggest an `endpoints` config?

```
authn:
  method: oidc
  oidc:
    issuer: ...
    audience: ...
      
endpoints:
  - openfga.v1.OpenFGAService/Write
     scopes:
       - openfga:write
  - openfga.v1.OpenFGAService/Read
     scopes:
       - openfga:read
```

Reply:
Yeah, it makes sense to split it out. I had it nested under the `oidc`/`preshared` block because the means of authorization differs between the two.

I suppose that the authorization block could be authn-type-agnostic, and we could enforce restrictions during config validation. For example, if you accidentally write a `keys` block when the `authn` method is `oidc`, the config schema could allow it but we can catch that in validation:

```
authz:
  global:
    scopes:
      - openfga:read
    keys:
      - cool-key-1 // specifying a key in addition to a scope would result in a config validation error
  endpoints:
```

Alternatively, if we want distinct schemas for each method type we could have blocks similar to `auth`:
```yaml
authz:
  oidc:
    global:
      scopes:
        - openfga:read
    endpoints:
      - openfga.v1.OpenFGAService/Write
           scopes:
             - openfga:write
  preshared:
    <empty here, but would be filled out when authn is preshared>
```